src/understand-tidb/table-statistics.md

```diff
@@ -1,6 +1,88 @@
-# Table statistics
+# TiDB Statistics
 
-* what statistics we are using, why
-* how the statics are collected
-* how the statics are used
+In the TiDB's SQL optimization phase, we will have the cases like which index of one table is most suitable one, whether a join operator should be a hash join, a index nested loop join or other join algorithm. Thus we need the statistics of the table to know its data distribution, guiding the TiDB choose the optimial execution plan.
 
+In this article, we would first introduce the data structure that database could use then have a easy guide of TiDB's statistics.
+
+## Statistics introduction
+
+There're many data structures to show the distribution of the data. Here we introduce three kind of them: histogram, count-min sketch and top-n value(most frequnt values).
+
+### Histogram
+
+#### Introduction
+
+Histogram splits the data into many buckets and uses some simple things to describing the bucket, such as how many records in in this bucket. It's widely used in many RDBMS to do the range estimation. We have two different type of histogram depending on the bucketing strategy: equal-depth histogram and equal-width histogram.
+
+We choose the equal-depth histogram according to the paper [Accurate estimation of the number of tuples satisfying a condition](https://dl.acm.org/citation.cfm?id=602294). The equal-depth histogram has a better guarantee of the error rate compared in the worst cases, compared with the equal-width histogram. The so-called equal-depth histogram means that the number of values falling into each bucket is as equal as possible. For example, we want to split the given records set `1.6, 1.9, 1.9, 2.0, 2.4, 2.6, 2.7, 2.7, 2.8, 2.9, 3.4, 3.5` into 4 buckets. Then we would get the final result as `[1.6, 1.9], [2.0, 2.6], [2.7, 2.8], [2.9, 3.5]`, the depth of each bucket is 3, i.e. the number of records in each bucket is 3. The graph is shown as below.
+
+![equal-depth histogram](/src/img/stats-histogram.png)
+
+#### Constrcution
+
+The construction will vary depending on whether we know the size of the complete dataset.
+
+If we know the size, we can dicide the depth of the bucket. So we just iterating each value `v`:
+
+- If `v` equals to the previous one, we just place it to the same bucket no matter the bucket is full or not. This can make sure that the same value is always in the same bucket.
+- If not, we check whether the bucket is full or not. If not full, put it into the current bucket. Otherwise we put it to a new bucket.
+
+If we don't know the size, we construct the histogram in the following way. We initialize the bucket depth to 1 for each bucket. And just inserting the data like before. Once the we meet the case that one bucket exceeds the needed depth, we double the depth of the bucket and combine two adjacent buckets into one bucket.
+
+### Count-Min Sketch(Legacy in TiDB)
```

Comment:
Why it becomes legacy?

Reply:
... and where is the related code about count-min sketch?

Reply:
@xxchan The explanation is written at the place where we introduce the top-n.

Reply:
@winoros so you can mention here briefly that it is superseded by top-n?